### ISSUE TITLE
InMemoryCache limits

### ISSUE THREAD

**Issue by sotomaque** [apollographql/apollo-client#7951]
Are there recommended cache size limits? Are these different per devices? What is the recommended approach for caching large query results?

Currently experiencing an app crash in react native when I cache a query (response includes 2000 objects).

**Comment by benjamn** [comment]
@sotomaque Having just done some profiling of React Native Android, I can tell you: even the newer Hermes JavaScript engine will struggle to write 2000 objects at once (at least in the Pixel 3 emulator). You can expect it to take on the order of ~500ms, though obviously the sizes of the objects matter. This same work takes (again very roughly) ~10-20ms on a desktop computer.

I'm less familiar with the memory limits imposed on React Native applications by Android and iOS, though I imagine memory may be pretty constrained, since it's the memory allocated to the JavaScript engine running within the application.

Could you tell what exactly caused the app crash? Was it consistent with the app hitting a memory limit and getting terminated by the OS?

Some general short-term solutions:
* You can skip the cache for unavoidably large queries by using `fetchPolicy: "no-cache"`, though the query result will still need to fit in memory.
* You can use [pagination](https://www.apollographql.com/docs/react/pagination/overview/) to reduce the initial size of large lists within the result tree, which should have a linear effect on cache write/read times.

**Comment by benjamn** [comment]
A few more questions:
1. What version of React Native are you using?
2. Which platform (iOS/Android) and device/emulator model number?
3. If you're on Android, are you using [Hermes](https://hermesengine.dev/) or the default JS engine?

**Comment by sotomaque** [comment]
this is occurring on an iphone simulator (using expo); changing the fetch policy to no-cache does prevent the app crash which is what leads me to believe its related to cache limit size. This is the react-native version I am currently using https://github.com/expo/react-native/archive/sdk-39.0.0.tar.gz; I have considered pagination but the point of this query is to grab a bulk amount of data; the objects being returned are really small, three fields, i.e something like { id: 1, name: '', email: '' }

**Comment by ajbouh** [comment]
I've seen nodejs crash for me as well when using `InMemoryCache`.

The top of the stacktrace also refers to `ObjectCanon.prototype.admit`. Here's a short except of that area of code:

``` javascript
  ObjectCanon2.prototype.admit = function(value) {
    var _this = this;
    if (isNonNullObject(value)) {
      var original = this.passes.get(value);
      if (original)
        return original;
      var proto = Object.getPrototypeOf(value);
      switch (proto) {
        case Array.prototype: {
          if (this.known.has(value))
            return value;
          var array = value.map(this.admit, this);
          var node2 = this.pool.lookupArray(array);
          if (!node2.array) {
            this.known.add(node2.array = array);
            if (__DEV__) {
              Object.freeze(array);
            }
          }
          return node2.array;
        }
        case null:
        case Object.prototype: {
          if (this.known.has(value))
            return value;
          var proto_1 = Object.getPrototypeOf(value);
          var array_1 = [proto_1];
          var keys = this.sortedKeys(value);
          array_1.push(keys.json);
          var firstValueIndex_1 = array_1.length;
          keys.sorted.forEach(function(key) {
            array_1.push(_this.admit(value[key]));
          });
          var node2 = this.pool.lookupArray(array_1);
          if (!node2.object) { // <-- this line is at the top of the stacktrace
            var obj_1 = node2.object = Object.create(proto_1);
            this.known.add(obj_1);
            keys.sorted.forEach(function(key, i) {
              obj_1[key] = array_1[firstValueIndex_1 + i];
```

The crash seemed to stop when I set `__DEV__` to false. Perhaps it's some interaction between WeakMap, garbage collection, and Object.freeze?

**Comment by phryneas** [comment]
@ajbouh the problem you were seeing here with `ObjectCanon.admit` is one of the potential memory problems we've tackled with the just-released Apollo Client 3.9 beta.

You (and everyone else following along) can read all the details in our [announcement blog post](https://www.apollographql.com/blog/apollo-3-9-beta-feature-spotlight-the-memory-story) :)

**Comment by phryneas** [comment]
We have recently released Apollo 3.9 which I believe should have fixed this issue - so I'm going to close this.

If this issue keeps persisting in Apollo Client >= 3.9.0, please open a new issue!
